### PR TITLE
[Draft] Local-Amax for Current-Scaling

### PR DESCRIPTION
**Pull request by mingxu1067**
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/cpp_extensions/quantization.py [hunk 1]
```diff
@@ -532,6 +537,134 @@ class QuantizePrimitive(BaseDBiasQuantizePrimitive):
...
 
 
+class AmaxCalculationPrimitive:
+    """
+    Amax Calculation Primitive with custom_partitioning
+    """
+
+    name = "jax_local_amax"
+    multiple_results = False
+    impl_static_args = (1, 2)  # use_global_amax
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/quantization.py]: Static arg names should be `# amax_across_tpsp, amax_across_fsdp`, right?

File: transformer_engine/jax/cpp_extensions/quantization.py [hunk 2]
```diff
@@ -532,6 +537,134 @@ class QuantizePrimitive(BaseDBiasQuantizePrimitive):
...
+                gmesh = global_mesh_resource()
+                amax = lax_paral_op(amax, jax.lax.pmax, gmesh.fsdp_resource, mesh)
+
+            return amax
+
+        arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
+        return mesh, sharded_impl, amax_sharding, arg_shardings
+
+
+def register_primitive_only_outter(cls):
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/quantization.py]: Self-resolve: `outter` -> `outer`

File: transformer_engine/jax/layernorm_mlp.py [hunk 3]
```diff
@@ -272,13 +272,12 @@ def _layernorm_mlp_fwd_rule(
         epsilon,
         norm_type,
         quantizer=ffn1_quantizer_set.x,
+        amax_across_tpsp=True,
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/layernorm_mlp.py]: Can you help me understand why `amax_across_tpsp` is being set to `True` in these cases? Is this because we will be all-gathering these GEMM inputs?

File: transformer_engine/jax/dense.py [hunk 4]
```diff
@@ -175,13 +183,15 @@ def _dense_fwd_rule(
...
         quantizer=quantizer_set.x,
+        amax_across_tpsp=using_global_amax_of_x,
     )
     casted_x = with_sharding_constraint_by_logical_axes(casted_x, input_axes)
 
     casted_kernel = tex.quantize(
         kernel,
         flatten_axis=flatten_axis_k,
         quantizer=quantizer_set.kernel,
+        amax_across_fsdp=True,
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/dense.py]: Same question here on why `amax_across_fsdp=True` here always? Is that because this will be All-Gathered in the GEMM for FSDP?

File: transformer_engine/jax/cpp_extensions/quantization.py [hunk 5]
```diff
@@ -532,6 +537,134 @@ class QuantizePrimitive(BaseDBiasQuantizePrimitive):
...
+    ):
+        """
+        amax calcuation abstract
+        """
+        del amax_across_tpsp, amax_across_fsdp
+
+        dtype = dtypes.canonicalize_dtype(x_aval.dtype)
+        assert dtype in [jnp.float32, jnp.float16, jnp.bfloat16]
+
+        out_aval = jax.core.ShapedArray(shape=(1,), dtype=jnp.float32)
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/quantization.py:571]: This is only storing a single-value as the global shape, so from JAX/XLA's perspective it will think this is the same value replicated across devices, rather than a unique value per device, so things could break easily if we have any resharding, right?

@phu0ngng last we synced I thought we were going with your idea for an amax tensor on each device that has the shape of the mesh. Did this plan change? Thanks!

Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/quantization.py:571]: Resolved per offline discussion on plan

File: transformer_engine/jax/cpp_extensions/quantization.py (hunk 2 above)
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/quantization.py]: Hi, may I know why we can't simply use the existing registration?

File: transformer_engine/jax/layernorm_mlp.py (hunk 3 above)
Comment (mingxu1067) [reply transformer_engine/jax/layernorm_mlp.py]: Yes, as the recipe suggested, we only need to run activation AR for those run AG right before GEMM, which is colmun-parallel layers in SP.

File: transformer_engine/jax/dense.py (hunk 4 above)
Comment (mingxu1067) [reply transformer_engine/jax/dense.py]: Yes, it is because we AG the weight then run the GEMM, therefore GEMM would need weight scale to unique one across all devices.

File: transformer_engine/jax/cpp_extensions/quantization.py (hunk 2 above)
Comment (mingxu1067) [reply transformer_engine/jax/cpp_extensions/quantization.py]: Makes sense. Added an arg to contol the scope of registration.

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 6]
```diff
@@ -1099,6 +1108,7 @@ def rmsnorm_fwd(
         zero_centered_gamma: If True, gamma is zero-centered.
         epsilon: Small constant for numerical stability.
         quantizer: Optional quantizer for FP8 quantization of the output.
+        amax_scope: Indicate the scope to run amax calculation. Default is AmaxScope.LOCAL.
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/normalization.py]: Can we expand this comment to indicate that this arg will have no effect if we are not using a form of current scaling?

File: transformer_engine/jax/dense.py [hunk 7]
```diff
@@ -238,6 +249,7 @@ def _dense_bwd_rule(
         is_dbias=use_bias,
         flatten_axis=flatten_axis_k,
         quantizer=quantizer_set.dgrad,
+        amax_scope=AmaxScope.TPSP,
```
Comment (phu0ngng) [inline transformer_engine/jax/dense.py]: Hi, 
I think we only do AG the `grad` only if the input in the FWD is not AG.
```suggestion
        amax_scope=AmaxScope.LOCAL if using_global_amax_of_x else AmaxScope.TPSP,
```

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 6 above)
Comment (mingxu1067) [reply transformer_engine/jax/cpp_extensions/normalization.py]: Done

File: transformer_engine/jax/dense.py (hunk 7 above)
Comment (mingxu1067) [reply transformer_engine/jax/dense.py]: Thank you for pointing out. Fixed.
